source/material_model/melt_boukare.cc

```diff
@@ -0,0 +1,1260 @@
+            {
+              out.thermal_expansion_coefficients[q] += (endmember_mole_fractions_in_composite[i] * endmembers.volumes[i] * endmembers.thermal_expansivities[i]) / total_volume;
+              out.specific_heat[q] += endmember_mole_fractions_in_composite[i] * endmembers.heat_capacities[i] / total_molar_mass;
+
+
+              if (endmember_states[i] == EndmemberState::solid && solid_molar_volume > 0)
+                {
+                  out.compressibilities[q] += (endmember_mole_fractions_in_composite[i] * endmembers.volumes[i])
+                                              / (solid_molar_volume * endmembers.bulk_moduli[i]);
+                }
+            }
+
+          if (solid_molar_volume > 0)
+            out.densities[q] = solid_molar_mass / solid_molar_volume;
+          else
+            out.densities[q] = melt_molar_mass / melt_molar_volume;
+
+          if (melt_out != nullptr)
+            {
+              double melt_compressiblity = 0;
+              for (unsigned int i=0; i<n_endmembers; ++i)
+                if (endmember_states[i] == EndmemberState::melt && melt_molar_volume > 0)
+                  melt_compressiblity += (endmember_mole_fractions_in_composite[i] * endmembers.volumes[i]) / (melt_molar_volume * endmembers.bulk_moduli[i]);
+
+              if (melt_molar_volume > 0)
+                melt_out->fluid_densities[q] = melt_molar_mass / melt_molar_volume;
+              else
+                {
+                  // make sure we have a useful melt density even if there is no melt to avoid density jumps
+                  double mass = 0, volume = 0;
+                  for (unsigned int i=0; i<n_endmembers; ++i)
+                    if (endmember_states[i] == EndmemberState::melt)
+                      {
+                        mass += endmember_mole_fractions_per_phase[i] * molar_masses[i];
+                        volume += endmember_mole_fractions_per_phase[i] * endmembers.volumes[i];
+                      }
+                  melt_out->fluid_densities[q] = mass/volume;
+                }
+
+              // This does not take into account the volume change due to thermal expansion of melt
+              melt_out->fluid_density_gradients[q] = melt_out->fluid_densities[q] * melt_out->fluid_densities[q]
+                                                     * melt_compressiblity
+                                                     * this->get_gravity_model().gravity_vector(in.position[q]);
+            }
+
+
+          if (this->include_melt_transport())
+            {
+              // Calculate the melting rate
+              const unsigned int Fe_solid_idx = this->introspection().compositional_index_for_name("molar_Fe_in_solid");
+              const unsigned int Fe_melt_idx = this->introspection().compositional_index_for_name("molar_Fe_in_melt");
+              const unsigned int porosity_idx = this->introspection().compositional_index_for_name("porosity");
```

Comment:
put these outside the loop at the top of the function (I think porosity_idx is already duplicated).